numhask/src/NumHask/Algebra/Group.hs

```diff
@@ -0,0 +1,117 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ConstrainedClassMethods #-}
+{-# LANGUAGE RoleAnnotations #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE MonoLocalBinds #-}
+-- | The Group hirarchy
+module Numhask.Algebra.Group
+      ( Magma(..)
+      , Unital(..)
+      , Semigroup(..)
+      , Commutative
+      , Absorbing(..)
+      , Invertible(..)
+      , Idempotent
+      , Monoid(..)
+      , Group
+      , groupSwap
+      , AbelianGroup
+      )
+      where
+
+import qualified Prelude                       as P
+
+-- * Magma structure
+-- | A <https://en.wikipedia.org/wiki/Magma_(algebra) Magma> is a tuple (T,comb) consisting of
+--
+-- - a type a, and
+--
+-- - a function (comb) :: T -> T -> T
+--
+-- The mathematical laws for a magma are:
+--
+-- - comb is defined for all possible pairs of type T, and
+--
+-- - comb is closed in the set of all possible values of type T
+--
+-- or, more tersly,
+--
+-- > ∀ a, b ∈ T: a comb b ∈ T
+--
+-- These laws are true by construction in haskell: the type signature of 'magma' and the above mathematical laws are synonyms.
+--
+--
+class Magma a where
+  comb :: a -> a -> a
+
+-- | A Unital Magma
+--
+-- > unit comb a = a
+-- > a comb unit = a
+--
+class Magma a =>
+      Unital a where
+  unit :: a
+
+-- | A semigroup is an associative Magma
+--
+-- > (a comb b) comb c = a comb (b comb c)
+class Magma a =>
+      Semigroup a where
+      infixl 6 <>
+      (<>) :: a -> a -> a
+      (<>) = comb
+
+-- | A Commutative Magma
+--
+-- > a comb b = b comb a
+class Magma a =>
+      Commutative a
+
+-- | A Monoid is a Semigroup with an identity element
+--
+class (Unital a, Semigroup a) => Monoid a where
+      mempty :: a
+      mempty = unit
+instance (Unital a, Semigroup a) => Monoid a
+
+-- | An Invertible Magma
+--
+-- > ∀ a ∈ T: a `comb` inv a = unit
+--
+class Unital a =>
+      Invertible a where
```

Comment:
How is Invertible different from Group?

Reply:
The thing is called a loop. It lacks the associativity.

I thought it would be better to define the invertible via the Unital because the law is so nice and easily testable, but looking at it now, I think the more general formulation as a weak inverse of the magma is also ok.
The law is (without the existence of an identity!): 
```
for all *a* exists an *a^(-1)* so that for all *b*:
a^-1 `comb` (a `comb` b) = b = (b `comb` a) `comb` a^-1
```